### ISSUE TITLE
trax logs of H5p activities do not get sent to LRS 

### ISSUE THREAD

**Issue by scizmeli** [trax-project/moodle-trax-logs#20]
Hi,
I configured trax on moodle 3.9.2. Trax is on "synchronous" mode. When a student interacts with H5p content, logs are written to moodle log stor, That part is OK.  Moodle cron works, it runs every minute. But no data get sent to LRS. I am pretty confident that we configured lrs settings (LRS endpoint, username and pass) correctly. 

Cron job always reports this, no matter we have new interactions or not :+1: 

`Execute scheduled task: Trax Logs: define courses (logstore_trax\task\define_courses_task)
... started 14:45:03. Current memory use 34.9MB.
... used 22 dbqueries
... used 0.066028118133545 seconds
Scheduled task complete: Trax Logs: define courses (logstore_trax\task\define_courses_task)
`

Also try to run unit tests :

/opt/bitnami/moodle# vendor/bin/phpunit admin/tool/log/store/trax/tests/store_test.php 
Moodle 3.9.2 (Build: 20200914)
Php: 7.3.24, mariadb: 11.2.3-MariaDB, OS: Linux 5.15.0-94-generic x86_64
PHPUnit 7.5.20 by Sebastian Bergmann and contributors.

EE                                                                  2 / 2 (100%)

Time: 9.54 seconds, Memory: 56.50 MB

There were 2 errors:

1) store_test::test_sync_process
array_merge() expects at least 1 parameter, 0 given

/bitnami/moodle/admin/tool/log/store/trax/classes/src/config.php:208
/bitnami/moodle/admin/tool/log/store/trax/classes/src/config.php:221
/bitnami/moodle/admin/tool/log/store/trax/tests/utils/base.php:138
/bitnami/moodle/admin/tool/log/store/trax/tests/utils/base.php:67
/bitnami/moodle/admin/tool/log/store/trax/tests/store_test.php:47
/bitnami/moodle/lib/phpunit/classes/advanced_testcase.php:80
phpvfscomposer:///bitnami/moodle/vendor/phpunit/phpunit/phpunit:60

To re-run:
 vendor/bin/phpunit "store_test" admin/tool/log/store/trax/tests/store_test.php

2) store_test::test_async_process
array_merge() expects at least 1 parameter, 0 given

/bitnami/moodle/admin/tool/log/store/trax/classes/src/config.php:208
/bitnami/moodle/admin/tool/log/store/trax/classes/src/config.php:221
/bitnami/moodle/admin/tool/log/store/trax/tests/utils/base.php:138
/bitnami/moodle/admin/tool/log/store/trax/tests/utils/base.php:67
/bitnami/moodle/admin/tool/log/store/trax/tests/store_test.php:81
/bitnami/moodle/lib/phpunit/classes/advanced_testcase.php:80
phpvfscomposer:///bitnami/moodle/vendor/phpunit/phpunit/phpunit:60

To re-run:
 vendor/bin/phpunit "store_test" admin/tool/log/store/trax/tests/store_test.php

ERRORS!
Tests: 2, Assertions: 0, Errors: 2.

I am a bit out of options now. Where am I doing wrong?
Thanks for your help

**Comment by sfraysse** [comment]
Hi,

> I am pretty confident that we configured lrs settings (LRS endpoint, username and pass) correctly.

Does your LRS receive other interactions than H5P, such as "logged-in" or "navigated-in"?
If not, you should start to solve this before going to H5P.

**Comment by scizmeli** [comment]
Thanks for the reply. Not I don't see any  "logged-in" or "navigated-in" data in the LRS.

Since I am new to this, I have no idea where to look for fixing this. Could you kindly give me a pointer? Where in moodle 3.9.2 should I look? 
thanks a lot beforehand for your kindly support

**Comment by sfraysse** [comment]
Hi,

First, be sure that your LRS is up and running and that the LRS API endpoint is valid, and works with the credentials you have. You don't need Moodle to do that. You can use a tool such as Postman if you know how the LRS API works.

Then, be sure to enter the endpoint and credentials correctly in Moodle. If this is not confidential, you can show me the LRS endpoint to check that there is no issue in its format.

Finally, you need to be sure that the LRS endpoint is accessible from your Moodle server, so that no security rule breaks the communication between both.

I hope this will help.

**Comment by scizmeli** [comment]
Our LRS (https://yetanalytics.github.io/lrsql/) is up and running. I can GET and POST using curl without problem. 

curl -X GET \
    https://lrsdev.melda.io/xapi/statements \
    -H "Content-Type: application/json" \
    -H "X-Experience-API-Version: 1.0.3" \
    -u "XXX:XXX"

Below is the screenshot of the trax config screen : 
![image](https://github.com/trax-project/moodle-trax-logs/assets/1478968/a4fa2468-354c-4d7b-97ff-377d8f1ffaed)

Both LRS and Moodle are running inside the docker containers. To test, I entered inside the moodle container and ran the above curl command (with the full domain name). I was able to get the result.

The problem seems somewhere else. I can't seem to get a clue why moodle/trax are failing to send stuff to the LRS... Cron logs are empty (to me it was weird that they show 0 for the push job). Also, as I stated above, for some reason I can't run unit tests. If I could, maybe that would give me more clues?

Meanwhile, interactions with H5P are logged in moodle.
![image](https://github.com/trax-project/moodle-trax-logs/assets/1478968/2445a919-c15d-4b69-98d0-e07ee47e3db3)

Any ideas?
Thanks

**Comment by sfraysse** [comment]
Hi,
The endpoint in the Moodle config should not end with `/statements`.
This is the general endpoint, not the statements API endpoint.

**Comment by scizmeli** [comment]
OMG thank you very much. I lost straight 2 weeks because of that. Really did not see it coming....

**Comment by scizmeli** [comment]
Btw are you familiar with the moodle 4.2 equivalent of the same stuff? 

 I tried logstore  + mod_hvp plugins a try but it seems the outcome is not as detailed as with trax... For example I do not see question answers and grading in the LRS.

Should I keep with trax or are the better alternatives that are known to work well for moodle 4.2... It is funny how there is so little to none recent tutorials on this topic.   

Again very kind of you to help out. A big merci beaucoup.

**Comment by sfraysse** [comment]
TRAX Logs should work with Moodle 4.2. I didn't test the H5P events but the other events work for sure.
I have no other alternative to recommend.

**Comment by scizmeli** [comment]
I didn't know. In the trax documentation I remember it is stated that it is compatible with 3.9... I'll check.

**Comment by sfraysse** [comment]
Yes, it's not official, but thanks to some PR, it should work with Moodle 4.2.
